sound/soc/sof/intel/hda-loader.c

```diff
@@ -177,13 +176,6 @@ static int cl_dsp_init(struct snd_sof_dev *sdev, int stream_tag)
 			__func__);
 
 err:
-	flags = SOF_DBG_DUMP_REGS | SOF_DBG_DUMP_PCI | SOF_DBG_DUMP_MBOX;
-
-	/* force error log level after max boot attempts */
-	if (hda->boot_iteration == HDA_FW_BOOT_ATTEMPTS)
-		flags |= SOF_DBG_DUMP_FORCE_ERR_LEVEL;
-
-	hda_dsp_dump(sdev, flags);
```

Comment:
wasn't this explicitly done before `snd_sof_dsp_core_power_down()` below because after the power down some information for the dump becomes inaccessible? Although ATM I cannot confirm that this indeed would be a problem. Also, this will now disable dumps for intermittent failures, fixed by a retry. Which I personally find good, but maybe others will want to keep at least a debug print for such failures

Reply:
> wasn't this explicitly done before `snd_sof_dsp_core_power_down()` below because after the power down some information for the dump becomes inaccessible? Although ATM I cannot confirm that this indeed would be a problem. 

The dsp dumps are parsing the already acquired oops and stack, it does not reach out to the DSP anymore, but yes, this is a valid concern.
For me eyes and tests it is safe and at the end we will have similarly structured debug information among platforms.

> Also, this will now disable dumps for intermittent failures, fixed by a retry. Which I personally find good, but maybe others will want to keep at least a debug print for such failures

Yes, it is a drawback for sure. I'm not sure how frequently we face with failing DSP boot which is fixed by a simple retry.
Adding a `dev_dbg()` if the DSP needs to be nagged to stay alive might be sufficient - if it did not booted for the first try.

Reply:
@ujfalusi I think this was something we changed recently to prevent reading all F's when cl_dsp_init failed. Are you sure we dont need this?

Reply:
@ranj063, we will print the full dump as previously after the last attempt failed but no prints for the re-tries.

If we want to reduce the amount of dumps, we need to remove some of the prints and in order to be able to control it they should not come from all over the place, but they should be under control.

It might be beneficial to move the re-try handling to core level and do a complete and full reset from scratch all the time (and we can do that for all platforms, not only for HDA). That way we can introduce a command line parameter or a new flag to tell the core to print for all failure a dump.
That can be done as an update to this PR?

Reply:
@ujfalusi but my main concern is after the last attempt, we would have powered off the DSP already no? there was an issue raised about this previously. Take a look at https://github.com/thesofproject/linux/issues/2805

Reply:
@ranj063, we would power off the host managed cores, yes. I think the DSP will remain powered still.
The last failure will trigger the snd_sof_dsp_dbg_dump() in core (laoder.c:snd_sof_run_firmware()) which invokes the sof_ops(sdev)->dbg_dump() and that is the same hda_dsp_dump() we called, but you (and @lyakh) are right that we will call it _after_ the host managed cores are powered down.
If I read the code right, the hda_dsp_dump() will read registers from the DSP_BAR (STATUS, FW_TARCEP) then it reads from mmio regions to get additional information, like xoops, panic_info and stack_words.

Will all of these going to be invalidated when we shut down the host managed cores of the DSP?

Would you be more comfortable if the snd_sof_dsp_dbg_dump() would be called from the platform codes and not from the core in case of the firmware fails to boot?
If I look more closely, this is only a case for HDA, which makes it a bit hard to handle in a nice way, but I'll try a dump before and after a host managed core off and see if they differ in any meaningful way (and also do two dumps while the cores are still enabled).

Reply:
> Will all of these going to be invalidated when we shut down the host managed cores of the DSP?
Yes, I think so. If you look at the dump before and after powering off the cores, you can see the difference.
> Would you be more comfortable if the snd_sof_dsp_dbg_dump() remained to be called from the platform codes and not from the core in case of the firmware fails to boot?
I think that's OK if thats the only way to get a meaningful dump

Reply:
> > Will all of these going to be invalidated when we shut down the host managed cores of the DSP?
> 
> Yes, I think so. If you look at the dump before and after powering off the cores, you can see the difference.

It does make a difference and it is not good.
 
> > Would you be more comfortable if the snd_sof_dsp_dbg_dump() remained to be called from the platform codes and not from the core in case of the firmware fails to boot?
> 
> I think that's OK if thats the only way to get a meaningful dump

Right, it looks like it is not as simple as I initially thought:
```
[ 1566.681635] sof-audio-pci-intel-tgl 0000:00:1f.3: error: cl_dsp_init: timeout HDA_DSP_SRAM_REG_ROM_STATUS read
[ 1566.681636] sof-audio-pci-intel-tgl 0000:00:1f.3: cl_dsp_init: dump 1:
[ 1566.681637] sof-audio-pci-intel-tgl 0000:00:1f.3: ------------[ DSP dump start ]------------
[ 1566.681646] sof-audio-pci-intel-tgl 0000:00:1f.3: error: extended rom status:  0x5000001 0x0 0x0 0x0 0x0 0x0 0x2530502 0x0
[ 1566.681646] sof-audio-pci-intel-tgl 0000:00:1f.3: ------------[ DSP dump end ]------------
[ 1566.681647] sof-audio-pci-intel-tgl 0000:00:1f.3: cl_dsp_init: dump 2:
[ 1566.691649] sof-audio-pci-intel-tgl 0000:00:1f.3: ------------[ DSP dump start ]------------
[ 1566.691659] sof-audio-pci-intel-tgl 0000:00:1f.3: error: extended rom status:  0x5000001 0x0 0x0 0x0 0x0 0x0 0x2530502 0x0
[ 1566.691660] sof-audio-pci-intel-tgl 0000:00:1f.3: ------------[ DSP dump end ]------------
[ 1566.691666] sof-audio-pci-intel-tgl 0000:00:1f.3: cl_dsp_init: dump 3:
[ 1566.701668] sof-audio-pci-intel-tgl 0000:00:1f.3: ------------[ DSP dump start ]------------
[ 1566.701677] sof-audio-pci-intel-tgl 0000:00:1f.3: error: extended rom status:  0xffffffff 0xffffffff 0xffffffff 0xffffffff 0xffffffff 0xffffffff 0xffffffff 0xffffffff                                                                                                 
[ 1566.701678] sof-audio-pci-intel-tgl 0000:00:1f.3: ------------[ DSP dump end ]------------
[ 1566.701679] sof-audio-pci-intel-tgl 0000:00:1f.3: error: dsp init failed after 3 attempts with err: -22
[ 1566.701697] sof-audio-pci-intel-tgl 0000:00:1f.3: error: failed to reset DSP
[ 1566.701698] sof-audio-pci-intel-tgl 0000:00:1f.3: ------------[ DSP dump start ]------------
[ 1566.701722] sof-audio-pci-intel-tgl 0000:00:1f.3: error: extended rom status:  0x1009701 0x1d003c 0x1 0x0 0x10000 0xc00 0x180030 0x0                                                                                                                                   
[ 1566.701723] sof-audio-pci-intel-tgl 0000:00:1f.3: ------------[ DSP dump end ]------------
[ 1566.701725] sof-audio-pci-intel-tgl 0000:00:1f.3: error: failed to boot DSP firmware -22
[ 1566.701796] hda_dsp_remove: ENTER
[ 1566.702760] sof-audio-pci-intel-tgl 0000:00:1f.3: error: sof_probe_work failed err: -22
```
I just set the ret to -EINVAL to simulate reset failure then dump twice (dump 1 / dump 2) followed by host managed cores power down and a third dump (dump 3). The third is completely useless.
However at the top level, from `snd_sof_run_firmware()` I also print the dump (failed to reset DSP) and that will be printing something which looks like a valid dump.

What is interesting is that if I skip the power down in `cl_dsp_init` then dump 3 is going to be same as dump 1/2 (surprise), but the dump from the top level is exactly what we have in the included log.

I also think that as a step 0 in `cl_dsp_init()` we should be powering down the host managed cores. We could as well remove the power down from the failure path, the `hda_dsp_remove()` will do that later on and it does not seam to have much affect when we leave the dsp_run...

I'll move this PR to draft for now.